### PR TITLE
Code refactor

### PR DESCRIPTION
**Pull request by joachimBurket**
Refactor of the code
* Moved all the code in `resources/lib`
* Main plugin code in `resources/lib/plugin.py`
* SRGSSR API clients in `resources/lib/srgssr_api_client/`
* Added French strings

### CODE REVIEW

**Comment by joachimBurket** [comment]
Hey, here's the v1 of the refactor! :tada: 
Just the following precisions:
* Not using the `DisableLetterMenu` for now as I'm thinking about changing the menus (#14)
* Not sure about the `socket.setdefaulttimeout(30)` method that was called at line 29 of `addon.py`. What does it do?
* Not sure about the `onlyActiveShows` param of the `get_tv_shows()`, and couldn't find it in the API documentation
* Getting episode title in the API response from `episodeList[i].mediaList[0].title` instead of `episodeList[i].title`, because in the RTS BU, it was the TVShow title most of the time
* There are some TODO in `resources/language/resource.language.de_de/strings.po`. If you can take a look, I think it would be better than a google tanslate :sweat_smile: :pray:

**Comment by ManBehindMooN** [comment]
I installed the add-on on my windows test machine and it did not work:
2022-04-19 11:22:10.019 T:16184   ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'json.decoder.JSONDecodeError'>
                                                   Error Contents: Expecting value: line 1 column 1 (char 0)
                                                   Traceback (most recent call last):
                                                     File "C:\Users\Man Behind Moon\AppData\Roaming\Kodi\addons\plugin.video.srgssr_ch_replay\addon.py", line 4, in <module>
                                                       plugin = Plugin()
                                                     File "C:\Users\Man Behind Moon\AppData\Roaming\Kodi\addons\plugin.video.srgssr_ch_replay\resources\lib\plugin.py", line 42, in __init__
                                                       self.video_client, self.subs_client = self._create_api_clients()
                                                     File "C:\Users\Man Behind Moon\AppData\Roaming\Kodi\addons\plugin.video.srgssr_ch_replay\resources\lib\plugin.py", line 60, in _create_api_clients
                                                       video_client = SRGSSRVideoApiClient(
                                                     File "C:\Users\Man Behind Moon\AppData\Roaming\Kodi\addons\plugin.video.srgssr_ch_replay\resources\lib\srgssr_api_client\srgssr_api_client.py", line 49, in __init__
                                                       self._auth_token = self.get_auth_token()
                                                     File "C:\Users\Man Behind Moon\AppData\Roaming\Kodi\addons\plugin.video.srgssr_ch_replay\resources\lib\srgssr_api_client\srgssr_api_client.py", line 83, in get_auth_token
                                                       return self.request_new_token()
                                                     File "C:\Users\Man Behind Moon\AppData\Roaming\Kodi\addons\plugin.video.srgssr_ch_replay\resources\lib\srgssr_api_client\srgssr_api_client.py", line 95, in request_new_token
                                                       data = res.json()
                                                     File "C:\Users\Man Behind Moon\AppData\Roaming\Kodi\addons\script.module.requests\lib\requests\models.py", line 900, in json
                                                       return complexjson.loads(self.text, **kwargs)
                                                     File "C:\Program Files\Kodi\system\python\Lib\json\__init__.py", line 357, in loads
                                                       return _default_decoder.decode(s)
                                                     File "C:\Program Files\Kodi\system\python\Lib\json\decoder.py", line 337, in decode
                                                       obj, end = self.raw_decode(s, idx=_w(s, 0).end())
                                                     File "C:\Program Files\Kodi\system\python\Lib\json\decoder.py", line 355, in raw_decode
                                                       raise JSONDecodeError("Expecting value", s, err.value) from None
                                                   json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
                                                   -->End of Python script error report<--

You might find the issue faster than I do...

File: plugin.video.srgssr_ch_replay/resources/settings.xml [hunk 1]
```diff
@@ -1,17 +1,90 @@
...
+					<level>0</level>
+					<default/>
+					<constraints>
+						<allowempty>true</allowempty>
+					</constraints>
+					<control format="string" type="edit">
+						<heading>30101</heading>
+					</control>
+				</setting>
+				<setting help="" id="consumer_secret" label="30102" type="string">
```
Comment (ManBehindMooN) [inline plugin.video.srgssr_ch_replay/resources/settings.xml:16]: The keys have changed so we need a migration temporary migration script for all these keys consumerKey -> consumer_key so the users do not have to re-input them. I suggest something like if consumer_key is empty but consumerKey is not then we set it into consumer_key the first when the add-on is started. We should do this for all the settings and the migration script can then be removed again

File: plugin.video.srgssr_ch_replay/addon.xml [hunk 2]
```diff
@@ -39,4 +39,4 @@ Um die Sendungen streamen zu können, muss ein Konto bei der SRG SSR erstellt we
     <website>https://www.srgssr.ch</website>
     <forum>https://github.com/ManBehindMooN/kodi_plugin_video_srgssr_ch_replay/discussions</forum>
   </extension>
-</addon>
\ No newline at end of file
+</addon>
```
Comment (ManBehindMooN) [inline plugin.video.srgssr_ch_replay/addon.xml:42]: The French translation for 
* <summary lang="fr_FR"> 
* <summary lang="fr_FR">
*  <disclaimer lang="fr_FR"> 

Should be included into the addon.xml

File: plugin.video.srgssr_ch_replay/resources/language/resource.language.de_de/strings.po [hunk 3]
```diff
@@ -40,13 +40,21 @@ msgctxt "#30022"
...
+
 msgctxt "#30098"
 msgid "In order to use this add-on a consumer key & secret in your add-on settings must be provided.[CR]See instructions: [B]https://github.com/ManBehindMooN/[CR]kodi_plugin_video_srgssr_ch_replay[/B][CR][CR]Sorry for the inconvenience."
 msgstr "Um das Add-on benutzen zu können braucht es Consumer Keys & Secrets in den Add-on Settings.[CR]Instruktionen unter: [B]https://github.com/ManBehindMooN/[CR]kodi_plugin_video_srgssr_ch_replay[/B][CR][CR]Sorry für die Umstände."
 
 msgctxt "#30099"
-msgid "Unexpected error"
-msgstr "Unerwarteter Fehler"
+msgid "API credentials missing"
+msgstr "TODO"
```
Comment (ManBehindMooN) [inline plugin.video.srgssr_ch_replay/resources/language/resource.language.de_de/strings.po:57]: I will do the German translation as soon as we have it merged

**Comment by ManBehindMooN** [comment]
Side note: As soon as I have a running add-on I will test if the Favorites are working and if they a backward compatible. If not it must be mentioned in the addon.xml as change.

**Comment by joachimBurket** [comment]
Hey, your welcome it was nice working on it :)

> onlyActiveShows => it is not documented I found it by reverse engineering the Play SRF website how they did it. See https://github.com/ManBehindMooN/kodi_plugin_video_srf_ch_replay/blob/2.0.3-fix%231/plugin.video.srf_ch_replay/addon.py line 31. There I keep track what the SRF Play website is doing...

Thanks for the info, I've kept the same logic with the setting.

> Do not worry about the German translation :)

Thanks ! 

> I installed the add-on on my windows test machine and it did not work:
> 2022-04-19 11:22:10.019 T:16184 ERROR : EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
> - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
> Error Type: <class 'json.decoder.JSONDecodeError'>
> Error Contents: Expecting value: line 1 column 1 (char 0)

That's strange. It seems that the response returned by the call to `https://api.srgssr.ch/oauth/v1/accesstoken?grant_type=client_credentials` is empty :thinking: 
I haven't encountered this behavior, so I'm not sure what it can be. Could you try to curl the URL with the Basic auth header and see if it passes?

> As soon as I have a running add-on I will test if the Favorites are working and if they a backward compatible. If not it must be mentioned in the addon.xml as change.

I tested it and I think they're not compatible unfortunately :/ 

> The Inputstream Adaptive should be mentioned in the README in order to help users who will experience a similar resolution problem as we did

Good call, I'll add something to the README

File: plugin.video.srgssr_ch_replay/addon.xml (hunk 2 above)
Comment (joachimBurket) [reply plugin.video.srgssr_ch_replay/addon.xml:42]: Oh forgot them! I'll also add them ;)

File: plugin.video.srgssr_ch_replay/resources/settings.xml (hunk 1 above)
Comment (joachimBurket) [reply plugin.video.srgssr_ch_replay/resources/settings.xml:16]: Didn't thought about that, that seems a good idea to me :+1:

**Comment by ManBehindMooN** [comment]
> I haven't encountered this behavior, so I'm not sure what it can be. Could you try to curl the URL with the Basic auth header and see if it passes?

Yeah, it is really strange. I will investigate.

**Comment by ManBehindMooN** [comment]
> > I haven't encountered this behavior, so I'm not sure what it can be. Could you try to curl the URL with the Basic auth header and see if it passes?
> 
> Yeah, it is really strange. I will investigate.

@joachimBurket I found the problem:

this works:
res = requests.post(self._base_url + "/oauth/v1/accesstoken",params=params, auth=self._basic_auth)
this does not work under Kodi windows:
res = requests.post(os.path.join(self._base_url, "oauth/v1/accesstoken"),params=params, auth=self._basic_auth)
==> os.path.join seems to be the problem

**Comment by ManBehindMooN** [comment]
> 



> 
And

File: plugin.video.srgssr_ch_replay/resources/lib/plugin.py [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+            tvshow_id = params.get("tvShowId")
+            current_page = int(params.get("currentPage", 1))
+            number_of_episodes = int(params.get("numberOfEpisodes", 0))
+            next_page_id = params.get("nextPageId", "")
+            self.list_episodes(bu, tvshow_id, current_page, number_of_episodes, next_page_id)
+        elif mode == "listTvShowsByLetter":
+            letter = params.get("letter")
+            self.list_tv_shows(bu, letter)
+        elif mode == "searchTvShows":
+            self.search_tv_shows()
```
Comment (ManBehindMooN) [inline plugin.video.srgssr_ch_replay/resources/lib/plugin.py:112]: search_tv_shows() is not implemented

**Comment by ManBehindMooN** [comment]
All in all it looks good besides the already mentioned issues or improvements.

File: plugin.video.srgssr_ch_replay/resources/lib/srgssr_api_client/srgssr_api_client.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        expires_in = data["expires_in"]
+        token_exp = (datetime.utcnow() + timedelta(seconds=expires_in)).isoformat()
+        self._logger.debug(f"Got a new AuthToken valid until {token_exp}")
+
+        setattr(self._plugin.settings, f"srgssr_{self.api_name}_token", token)
+        setattr(self._plugin.settings, f"srgssr_{self.api_name}_token_exp", token_exp)
+
+        return token
+
+    def _renew_access_token(func):
```
Comment (ManBehindMooN) [inline plugin.video.srgssr_ch_replay/resources/lib/srgssr_api_client/srgssr_api_client.py:106]: got 4 PyLint issues here (rest of the code is good)
image

File: plugin.video.srgssr_ch_replay/resources/language/resource.language.de_de/strings.po (hunk 3 above)
Comment (ManBehindMooN) [reply plugin.video.srgssr_ch_replay/resources/language/resource.language.de_de/strings.po:57]: msgctxt "#30096"
msgid "Incorrect '{0} API' credentials"
msgstr "Falsche '{0} API' Zugangsdaten"

msgctxt "#30097"
msgid "Please check they are correctly entered in the add-on settings"
msgstr "Bitte sicherstellen, dass die Zugangsdaten korrekt in den Add-on Einstellungen eingegeben wurden"

msgctxt "#30099"
msgid "API credentials missing"
msgstr "API Zugangsdaten fehlen"

File: plugin.video.srgssr_ch_replay/resources/language/resource.language.en_gb/strings.po [hunk 6]
```diff
@@ -40,12 +40,21 @@ msgctxt "#30022"
 msgid "Enter a TV show name to search for"
 msgstr ""
 
+msgctxt "#30096"
+msgid "Incorrect '{0} API' credentials"
+msgstr ""
+
+msgctxt "#30097"
+msgid "Please check they are correctly entered in the addon settings"
```
Comment (ManBehindMooN) [inline plugin.video.srgssr_ch_replay/resources/language/resource.language.en_gb/strings.po:48]: please change to add-on instead of addon

**Comment by joachimBurket** [comment]
> this works:
> res = requests.post(self._base_url + "/oauth/v1/accesstoken",params=params, auth=self._basic_auth)
> this does not work under Kodi windows:
> res = requests.post(os.path.join(self._base_url, "oauth/v1/accesstoken"),params=params, auth=self._basic_auth)
> ==> os.path.join seems to be the problem for the _base_url
> ==> I guess urlparse.urljoin needs to be taken`

Good catch! I forgot that some os methods weren't working on any OS! I'll change them

File: plugin.video.srgssr_ch_replay/resources/lib/plugin.py (hunk 4 above)
Comment (joachimBurket) [reply plugin.video.srgssr_ch_replay/resources/lib/plugin.py:112]: I'm impementing it in the https://github.com/joachimBurket/kodi_plugin_video_srgssr_ch_replay/tree/feature/main_menu branch. I'll merge it as soon as it is running

**Comment by joachimBurket** [comment]
As I had to refactor a bunch of code for the main menu (in the https://github.com/joachimBurket/kodi_plugin_video_srgssr_ch_replay/tree/feature/main_menu branch), I'm doing the reviews on the main python files directly in this branch and will merge it in feature/3.0.4 when it's working.

**Comment by ManBehindMooN** [comment]
Hi, how are you doing? What is the time window for the remaining open issues/improvements and the like?

**Comment by joachimBurket** [comment]
Hi! I'm doing fine, sorry for the long delay, I was on holidays the last two weeks, so I was pretty deconnected.
I'm gonna work on it again next week. The blocking part was to be able to have enough reusable method to avoid repeating code for each new menu.
I will try to have a working main menu with the "shows by letter", "all shows", "search show" and "shows by topics" for the end of the next week. 
I'll keep you updated

**Comment by ManBehindMooN** [comment]
> Hi! I'm doing fine, sorry for the long delay, I was on holidays the last two weeks, so I was pretty deconnected. I'm gonna work on it again next week. The blocking part was to be able to have enough reusable method to avoid repeating code for each new menu. I will try to have a working main menu with the "shows by letter", "all shows", "search show" and "shows by topics" for the end of the next week. I'll keep you updated

Good to hear from you. I hope you had relaxing and nice holidays. Don't stress yourself there is no hurry I was just wondering :) Have a nice weekend!

**Comment by joachimBurket** [comment]
Hi, I had very relaxing holidays yes :)

Here are the last changes I worked on:
* A default BU can be set in the settings, skipping this menu in the addon
* The BU have their logos
* A main menu
* Latest videos sorted by topics
* Trending videos
* Search tv shows and videos
* The duration is now formatted Hour:Minutes:Seconds or Minutes:Seconds
* os.path error on Windows should now be ok

Could you test it and give me a feedback? 
 
PS: The new texts are hardcoded for now. I'll move them in the strings.po after your review

**Comment by ManBehindMooN** [comment]
Thank you for your great effort. I will have a thorough review over the next couple of days.

File: plugin.video.srgssr_ch_replay/resources/lib/srgssr_api_client/srgssr_api_client.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
+"""SRGSSR Base API Client"""
+
+import os
```
Comment (ManBehindMooN) [inline plugin.video.srgssr_ch_replay/resources/lib/srgssr_api_client/srgssr_api_client.py:3]: import os not used

File: plugin.video.srgssr_ch_replay/resources/lib/router.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
+"""Routing in the plugin menus"""
+
+import os.path
```
Comment (ManBehindMooN) [inline plugin.video.srgssr_ch_replay/resources/lib/router.py:3]: import os.path not used

Comment (ManBehindMooN) [reply plugin.video.srgssr_ch_replay/resources/lib/router.py:3]: and lib/__init__.py is empty...

**Comment by ManBehindMooN** [comment]
Feedback time :) In general the add-on run quiet good. You did some great work. There are only 3 issues in opinion:
* The settings keys are not migrated yet
* The subtitles do not work anymore
* TV Show Thumbnails for SRF are different to v3.0.3 (see list_episodes function > line 134:  picture = media.get('imageUrl')). In 3.0.4 the TV show thumbnail looks always the same. I guess we need to load the TV Show thumbnail differently by business unit.

A really nice option you have included is to choose a default business unit. So the only thing a I would miss in the future to choose a default main menu option for different options (All TV Shows, TV Shows by Letters...). But as discussed in https://github.com/ManBehindMooN/kodi_plugin_video_srgssr_ch_replay/issues/14 this should be added in the next version :)

From my point of view if the above 3 issues are fixed then we should deliver the new version.

**Comment by ManBehindMooN** [comment]
> Feedback time :) In general the add-on run quiet good. You did some great work. There are only 3 issues in opinion:
> 
>     * The settings keys are not migrated yet
> 
>     * The subtitles do not work anymore
> 
>     * TV Show Thumbnails for SRF & RTR & SwissInfo are different to v3.0.3 (see list_episodes function > line 134:  picture = media.get('imageUrl')). In 3.0.4 the TV show thumbnail looks always the same. RTS & RSI look fine. I guess we need to load the TV Show thumbnail differently by business unit.
> 
> 
> A really nice option you have included is to choose a default business unit. So the only thing a I would miss in the future to choose a default main menu option for different options (All TV Shows, TV Shows by Letters...). But as discussed in #14 this should be added in the next version :)
> 
> From my point of view if the above 3 issues are fixed then we should deliver the new version.

As an alternative instead of migrating you could just use the 4 original keys & secrets (consumerKey etc) and then no migration is needed. What do you think?

File: plugin.video.srgssr_ch_replay/resources/lib/router.py (hunk 8 above)
Comment (joachimBurket) [reply plugin.video.srgssr_ch_replay/resources/lib/router.py]: if i'm not mistaking, the `__init__.py` is needed to make a python package from a directory and then import `from resources.lib.plugin import Plugin` in the `addon.py` file.

**Comment by joachimBurket** [comment]
Thanks for your feedback! I'm having a look at the subtitles and the missing tv shows thumbnail. 

As for the keys in the settings, it's a great idea, I'll use the original keys, it will be simpler.

**Comment by joachimBurket** [comment]
About the subtitles, not sure if it's not a bug on SRGSSR API's side.

I did the following test on Postman:
* Get the latest episodes of the SRF's DOK TV Show (id 3b016ffc-afa2-466d-a694-c48b7ffe1783)
* Search episode with "subtitlesAvailable" at true
* Copy its episodeId (e.g. 6987ec59-471d-40e4-9a21-467a846ac3f6 for "7 Tage ... unter Männern")
* Get the subtitles (e.g. on `https://api.srgssr.ch/srgssr-play-subtitles/v1/identifier/urn:srf:episode:tv:6987ec59-471d-40e4-9a21-467a846ac3f6`)
  which returns the following answer on the three i've tested:
  ```json
  {
    "data": {
        "assets": [
            null
        ]
    }
  }
  ```

**Comment by joachimBurket** [comment]
If the new menu's titles seems good to you, i'll create the strings.po, and after that I think we're good to publish the new version! :)

**Comment by ManBehindMooN** [comment]
> Thanks for your feedback! I'm having a look at the subtitles and the missing tv shows thumbnail.
> 
> As for the keys in the settings, it's a great idea, I'll use the original keys, it will be simpler.

I'm gonna have a quick regression test and give a a feedback :)

**Comment by ManBehindMooN** [comment]
> About the subtitles, not sure if it's not a bug on SRGSSR API's side.
> 
> I did the following test on Postman:
> 
>     * Get the latest episodes of the SRF's DOK TV Show (id 3b016ffc-afa2-466d-a694-c48b7ffe1783)
> 
>     * Search episode with "subtitlesAvailable" at true
> 
>     * Copy its episodeId (e.g. 6987ec59-471d-40e4-9a21-467a846ac3f6 for "7 Tage ... unter Männern")
> 
>     * Get the subtitles (e.g. on `https://api.srgssr.ch/srgssr-play-subtitles/v1/identifier/urn:srf:episode:tv:6987ec59-471d-40e4-9a21-467a846ac3f6`)
>       which returns the following answer on the three i've tested:
>       ```json
>       {
>         "data": {
>             "assets": [
>                 null
>             ]
>         }
>       }
>       ```

The subtitles seem to work again. I guess it was a temporary problem on the SRG SSR side. I'm going to have a closer look at it. Thanks anyway for looking into it as well.

**Comment by ManBehindMooN** [comment]
Hi @joachimBurket, it looks good from my side. Do you wanna add anything or squash whatsoever? Just let me know then I'm going to merge it. What will happen after the merge.
1) I'm and Frank are going to test it on all our devices. Takes around a week. You can re-test it as well ;)
2) I'm going to adapt the readme and release notes
3) I'm going to deliver the version to the official repo

**Comment by joachimBurket** [comment]
Perfect, I just added the new menus in the strings.po, could you translate the "TODO" ones in the german strings? After that I think we're good to go!

**Pull request merged**